File: mcp-client-bridge/src/local/stdio-json-rpc.ts
/**
 * JSON-RPC communication over stdio pipes.
 * Handles message framing, request/response matching, and timeouts.
 */

import { ChildProcess } from 'node:child_process';

interface PendingRequest {
  resolve: (v: unknown) => void;
  reject: (e: Error) => void;
  timer: ReturnType<typeof setTimeout>;
}

export class StdioJsonRpc {
  private requestId = 0;
  private pendingRequests = new Map<number, PendingRequest>();
  private buffer = '';
  private process: ChildProcess | null = null;

  /** Attach to a child process's stdio streams. */
  attach(proc: ChildProcess): void {
    this.process = proc;
    proc.stdout?.on('data', (chunk: Buffer) => this.onData(chunk));
  }

  /** Detach and reject all pending requests. */
  detach(): void {
    this.rejectAll('Process detached');
    this.process = null;
    this.buffer = '';
  }

  /** Send a JSON-RPC request and wait for response. */
  sendRequest(method: string, params?: unknown, timeoutMs = 30_000): Promise<unknown> {
    return new Promise((resolve, reject) => {
      if (!this.process?.stdin?.writable) {
        reject(new Error('stdin not writable'));
        return;
      }
      const id = ++this.requestId;
      const msg = JSON.stringify({
        jsonrpc: '2.0',
        id,
        method,
        ...(params !== undefined && { params }),
      });

      const timer = setTimeout(() => {
        this.pendingRequests.delete(id);
        reject(new Error(`Request '${method}' timed out (${timeoutMs}ms)`));
      }, timeoutMs);

      this.pendingRequests.set(id, { resolve, reject, timer });
      this.process.stdin.write(msg + '\n');
    });
  }

  /** Send a JSON-RPC notification (no response expected). */
  sendNotification(method: string, params?: unknown): void {
    if (!this.process?.stdin?.writable) return;
    const msg = JSON.stringify({
      jsonrpc: '2.0',
      method,
      ...(params !== undefined && { params }),
    });
    this.process.stdin.write(msg + '\n');
  }

  /** Reject all pending requests with an error message. */
  rejectAll(reason: string): void {
    for (const [id, pending] of this.pendingRequests) {
      clearTimeout(pending.timer);
      pending.reject(new Error(reason));
      this.pendingRequests.delete(id);
    }
  }

  private onData(chunk: Buffer): void {
    this.buffer += chunk.toString();
    const lines = this.buffer.split('\n');
    this.buffer = lines.pop() ?? '';

    for (const line of lines) {
      if (!line.trim()) continue;
      try {
        const msg = JSON.parse(line);
        this.handleMessage(msg);
      } catch {
        // Non-JSON output, ignore
      }
    }
  }

  private handleMessage(msg: { id?: number; result?: unknown; error?: unknown }): void {
    if (msg.id == null) return; // Notification, ignore
    const pending = this.pendingRequests.get(msg.id);
    if (!pending) return;

    this.pendingRequests.delete(msg.id);
    clearTimeout(pending.timer);

    if (msg.error) {
      pending.reject(new Error(JSON.stringify(msg.error)));
    } else {
      pending.resolve(msg.result);
    }
  }
}

File: mcp-bridge-python/src/mcp_bridge/local/stdio_json_rpc.py
"""JSON-RPC communication over stdio pipes with asyncio."""

from __future__ import annotations

import asyncio
import json
import sys
from typing import Any


class StdioJsonRpc:
    """Handles JSON-RPC message framing, request/response matching, and timeouts."""

    def __init__(self) -> None:
        self._request_id = 0
        self._pending: dict[int, asyncio.Future[Any]] = {}
        self._reader_task: asyncio.Task[None] | None = None
        self._stdin: asyncio.StreamWriter | None = None
        self._stdout: asyncio.StreamReader | None = None

    def attach(self, stdin: asyncio.StreamWriter, stdout: asyncio.StreamReader) -> None:
        """Attach to a subprocess's stdio streams and start reading."""
        self._stdin = stdin
        self._stdout = stdout
        self._reader_task = asyncio.create_task(self._read_loop())

    def detach(self) -> None:
        """Detach and reject all pending requests."""
        if self._reader_task and not self._reader_task.done():
            self._reader_task.cancel()
        self._reader_task = None
        self._stdin = None
        self._stdout = None
        self.reject_all("Process detached")

    async def send_request(
        self, method: str, params: Any = None, timeout_ms: int = 30_000
    ) -> Any:
        """Send a JSON-RPC request and wait for response."""
        if not self._stdin:
            raise RuntimeError("stdin not attached")

        self._request_id += 1
        req_id = self._request_id
        msg: dict[str, Any] = {"jsonrpc": "2.0", "id": req_id, "method": method}
        if params is not None:
            msg["params"] = params

        future: asyncio.Future[Any] = asyncio.get_event_loop().create_future()
        self._pending[req_id] = future

        data = json.dumps(msg) + "\n"
        self._stdin.write(data.encode())
        await self._stdin.drain()

        try:
            return await asyncio.wait_for(future, timeout=timeout_ms / 1000)
        except asyncio.TimeoutError:
            self._pending.pop(req_id, None)
            raise TimeoutError(f"Request '{method}' timed out ({timeout_ms}ms)")

    def send_notification(self, method: str, params: Any = None) -> None:
        """Send a JSON-RPC notification (no response expected)."""
        if not self._stdin:
            return
        msg: dict[str, Any] = {"jsonrpc": "2.0", "method": method}
        if params is not None:
            msg["params"] = params
        data = json.dumps(msg) + "\n"
        self._stdin.write(data.encode())

    def reject_all(self, reason: str) -> None:
        """Reject all pending requests with an error."""
        for future in self._pending.values():
            if not future.done():
                future.set_exception(RuntimeError(reason))
        self._pending.clear()

    async def _read_loop(self) -> None:
        """Read lines from stdout and dispatch responses."""
        assert self._stdout is not None
        try:
            while True:
                line = await self._stdout.readline()
                if not line:
                    break
                self._handle_line(line.decode().strip())
        except asyncio.CancelledError:
            pass
        except Exception as e:
            self._log(f"Read loop error: {e}")

    def _handle_line(self, line: str) -> None:
        """Parse a JSON-RPC response line and resolve the pending future."""
        if not line:
            return
        try:
            msg = json.loads(line)
        except json.JSONDecodeError:
            return

        req_id = msg.get("id")
        if req_id is None:
            return

        future = self._pending.pop(req_id, None)
        if not future or future.done():
            return

        if "error" in msg:
            future.set_exception(RuntimeError(json.dumps(msg["error"])))
        else:
            future.set_result(msg.get("result"))

    @staticmethod
    def _log(msg: str) -> None:
        print(f"[stdio-rpc] {msg}", file=sys.stderr, flush=True)
